enterprise/reporting/flusher/flusher.go

```diff
@@ -0,0 +1,397 @@
+package flusher
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+	"github.com/rudderlabs/rudder-server/enterprise/reporting/flusher/aggregator"
+	"github.com/rudderlabs/rudder-server/enterprise/reporting/flusher/db"
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+)
+
+var (
+	flusherInstances = make(map[string]*Flusher)
+	flusherMu        sync.Mutex
+)
+
+type Flusher struct {
+	log logger.Logger
+
+	db                 db.DB
+	maxOpenConnections int
+
+	client     *http.Client
+	aggregator aggregator.Aggregator
+
+	instanceId string
+	table      string
+	module     string
+
+	sleepInterval                       config.ValueLoader[time.Duration]
+	flushWindow                         config.ValueLoader[time.Duration]
+	recentExclusionWindow               config.ValueLoader[time.Duration]
+	batchSizeFromDB                     config.ValueLoader[int]
+	aggressiveFlushEnabled              config.ValueLoader[bool]
+	lagThresholdForAggresiveFlushInMins config.ValueLoader[time.Duration]
+
+	reportingURL          string
+	minConcurrentRequests config.ValueLoader[int]
+	maxConcurrentRequests config.ValueLoader[int]
+	batchSizeToReporting  config.ValueLoader[int]
+
+	stats                   stats.Stats
+	minReportedAtQueryTimer stats.Measurement
+	reportsCounter          stats.Measurement
+	aggReportsTimer         stats.Measurement
+	aggReportsCounter       stats.Measurement
+	sendReportsTimer        stats.Measurement
+	deleteReportsTimer      stats.Measurement
+	concurrentRequests      stats.Measurement
+	reqLatency              stats.Measurement
+	reqCount                stats.Measurement
+	flushLag                stats.Measurement
+
+	commonTags stats.Tags
+}
+
+func NewFlusher(db db.DB, log logger.Logger, stats stats.Stats, conf *config.Config, table, reportingURL string, aggregator aggregator.Aggregator, module string) *Flusher {
+	flusherMu.Lock()
+	defer flusherMu.Unlock()
+
+	if instance, exists := flusherInstances[table]; exists {
+		return instance
+	}
```

Comment:
```suggestion
```

I would expect the factory to be responsible for doing this. It will also simplify your tests.